agent/consul/state/catalog.go

```diff
@@ -75,6 +75,9 @@ func gatewayServicesTableNameSchema() *memdb.TableSchema {
 						&ServiceIDIndex{
 							Field: "Service",
 						},
+						&memdb.IntFieldIndex{
+							Field: "Port",
```

Comment:
Important call out: We add `Port` as another member of this compound index, thus expanding the uniqueness constraint to the tuple `(gateway, service, port)`.

This allows us to not do complex updates to `GatewayService`s when config entries and services are added/removed.

Reply:
Wont this break some uses for terminating gateways. It wont prevent compilation but now to look things up in this index it will require 3 arguments to the `First` invocations as well as other methods on the `Txn`.

cc @freddygv